float64[] r
float64[] xm